File: app/src/main/java/com/falcon/restaurants/screens/common/BaseService.java
package com.falcon.restaurants.screens.common;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import androidx.annotation.Nullable;
import com.falcon.restaurants.common.MyApplication;
import com.falcon.restaurants.common.di.app.AppComponent;
import com.falcon.restaurants.common.di.service.ServiceComponent;
import com.falcon.restaurants.common.di.service.ServiceModule;

public class BaseService extends Service {

    private AppComponent getAppComponent() {
        return ((MyApplication)getApplication()).getAppComponent();
    }

    private ServiceComponent serviceComponent;
    public ServiceComponent getServiceComponent() {
        if(serviceComponent == null) {
            serviceComponent = getAppComponent().newServiceComponent(new ServiceModule(this));
        }
        return serviceComponent;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/falcon/restaurants/common/service/BaseService.kt
package com.falcon.restaurants.common.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import androidx.annotation.Nullable;
import com.falcon.restaurants.common.MyApplication;
import com.falcon.restaurants.common.di.app.AppComponent;
import com.falcon.restaurants.common.di.service.ServiceComponent;
import com.falcon.restaurants.common.di.service.ServiceModule;

class BaseService : Service() {

    val appComponent get() = (getApplication() as MyApplication).appComponent

    val serviceComponent by lazy {
        appComponent.newServiceComponent(ServiceModule(this))
    }

    // TODO to be changed
    override fun onBind(intent: Intent) : IBinder? {
        return null
    }
}
